students/DrewSmith/lessons/lesson02/assignment/code/charges_calc.py

```diff
@@ -0,0 +1,159 @@
+                    f"{str(json_decode_error)}")
+
+    except FileNotFoundError as file_not_found:
+        print(f"Source file not found: {filename}")
+        logging.error(
+            f"Source file not found: {filename}: {str(file_not_found)}")
+
+    return result
+
+def calculate_additional_fields(input_data):
+    '''
+    Calculates additional field values based on source file values
+
+    :param data: dictionary of source file values
+    '''
+    key_log = "Invalid data: {0}: missing {1} value: {2}"
+    for key, value in input_data.items():
+        try:
+            rental_start = datetime.datetime.strptime(
+                value['rental_start'], '%m/%d/%y')
+            rental_end = datetime.datetime.strptime(
+                value['rental_end'], '%m/%d/%y')            
+        except KeyError as key_err:
+            logging.warn(key_log.format(key, key_err.args[0], str(key_err)))
+            print(f"WARNING: {key} missing {key_err.args[0]} value")
+        except ValueError as value_error:
+            logging.warn(f"Invalid data: {key}: {str(value_error)}")
+            print(f"WARNING: {key} has an invalid rental date value")
+
+        value['total_days'] = (rental_end - rental_start).days
+
+        if value['total_days'] < 0:
+            logging.error(f"Invalid data: {key} rental start " + \
+                f"{rental_start} is greater than the end {rental_end}")
+            print(f"ERROR: {key} rental start is greater than end")
+
+        try:
+            value['total_price'] = value['total_days'] * value['price_per_day']
+        except KeyError as key_err:
+            logging.warn(key_log.format(key, key_err.args[0], str(key_err)))
+            print(f"WARNING: {key} missing {key_err.args[0]} value")
+        except ValueError as value_error:
+            logging.warn(f"Invalid data: {key}: {str(value_error)}")
+            print(f"WARNING: {key} has invalid price_per_day")
+
+        try:
+            value['sqrt_total_price'] = math.sqrt(value['total_price'])
+        except ValueError as value_error:
+            logging.warn(f"Invalid data: {key}: {str(value_error)}")
+            print(
+                f"WARNING: {key} unable square root the negative total_price")           
+        
+        try:
+            value['unit_cost'] = value['total_price'] / value['units_rented']
+        except KeyError as key_err:
+            logging.warn(key_log.format(key, key_err.args[0], str(key_err)))
+            print(f"WARNING: {key} missing {key_err.args[0]} value")
+        except ValueError as value_error:
+            logging.warn(f"Invalid data: {key}: {str(value_error)}")
+            print(f"WARNING: {key} has invalid units_rented value")
+        except ZeroDivisionError as zero_error:
+            logging.warn(
+                f"Invalid data: {key}: units_rented is 0: {str(zero_error)}")
+            print(f"WARNING: {key} units_rented cannot be 0")
```

Comment:
Very nice. It is a lot of try/except blocks, but so it goes with sanitizing data.